File: java/src/main/java/io/aurigraph/dlt/sdk/namespace/GraphQLApi.java
package io.aurigraph.dlt.sdk.namespace;

import io.aurigraph.dlt.sdk.AurigraphClient;
import io.aurigraph.dlt.sdk.dto.GraphQLResponse;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Namespace for /api/v11/graphql — GraphQL gateway.
 */
public class GraphQLApi {

    private final AurigraphClient client;

    public GraphQLApi(AurigraphClient client) {
        this.client = client;
    }

    /**
     * Execute an arbitrary GraphQL query with optional variables.
     */
    public GraphQLResponse query(String query, Map<String, Object> variables) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("query", query);
        if (variables != null && !variables.isEmpty()) {
            body.put("variables", variables);
        }
        return client.post("/graphql", body, GraphQLResponse.class);
    }

    /**
     * Execute a GraphQL query without variables.
     */
    public GraphQLResponse query(String query) {
        return query(query, null);
    }

    /**
     * Query all channels.
     */
    public GraphQLResponse queryChannels() {
        return query("{ channels { id name type status stakeholders { id role } } }");
    }

    /**
     * Query assets for a specific channel.
     */
    public GraphQLResponse queryAssets(String channelId) {
        Map<String, Object> vars = Map.of("channelId", channelId);
        return query("query($channelId: ID!) { assets(channelId: $channelId) { id type status owner } }", vars);
    }

    /**
     * Query all contracts.
     */
    public GraphQLResponse queryContracts() {
        return query("{ contracts { id title status templateId parties { name role } } }");
    }

    /**
     * Query node metrics.
     */
    public GraphQLResponse queryNodeMetrics() {
        return query("{ nodeMetrics { totalNodes activeNodes validatorCount networkStatus } }");
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/apis/GraphQLApi.kt
package io.aurigraph.sdk.apis

import io.aurigraph.sdk.AurigraphClient
import io.aurigraph.sdk.models.GraphQLRequest
import io.aurigraph.sdk.models.GraphQLResponse
import kotlinx.serialization.json.JsonObject

/**
 * GraphQL query interface for flexible data fetching.
 */
class GraphQLApi(private val client: AurigraphClient) {

    /** Execute an arbitrary GraphQL query. */
    suspend fun query(query: String, variables: JsonObject? = null): GraphQLResponse =
        client.post("/graphql", GraphQLRequest(query, variables))

    /** Convenience: query channels via GraphQL. */
    suspend fun queryChannels(fields: String = "id name channelType status"): GraphQLResponse =
        query("{ channels { $fields } }")

    /** Convenience: query assets via GraphQL. */
    suspend fun queryAssets(
        useCaseId: String? = null,
        fields: String = "id name assetType status",
    ): GraphQLResponse {
        val filter = useCaseId?.let { "(useCaseId: \"$it\")" } ?: ""
        return query("{ assets$filter { $fields } }")
    }

    /** Convenience: query contracts via GraphQL. */
    suspend fun queryContracts(fields: String = "id templateId status deployedAt"): GraphQLResponse =
        query("{ contracts { $fields } }")

    /** Convenience: query node metrics via GraphQL. */
    suspend fun queryNodeMetrics(fields: String = "totalNodes activeNodes validatorCount networkStatus"): GraphQLResponse =
        query("{ nodeMetrics { $fields } }")
}
